subsys/mgmt/mcumgr/smp_udp.c

```diff
@@ -122,26 +123,33 @@ static void smp_udp_receive_thread(void *p1, void *p2, void *p3)
 		struct sockaddr addr;
 		socklen_t addr_len = sizeof(addr);
 
-		int len = recvfrom(conf->sock, conf->recv_buffer,
-				   CONFIG_MCUMGR_SMP_UDP_MTU,
-				   0, &addr, &addr_len);
+		if (!nb) {
+			nb = mcumgr_buf_alloc();
+		}
+
+		int len = recvfrom(conf->sock,
+				   nb ? (char *)net_buf_simple_tail(&nb->b) : conf->recv_buffer,
+				   CONFIG_MCUMGR_SMP_UDP_MTU, 0, &addr, &addr_len);
 
 		if (len > 0) {
 			struct sockaddr *ud;
-			struct net_buf *nb;
-
 			/* store sender address in user data for reply */
-			nb = mcumgr_buf_alloc();
-			if (!nb) {
-				LOG_ERR("Failed to allocate mcumgr buffer");
-				/* No free space, drop smp frame */
-				continue;
+			if (nb) {
+				net_buf_simple_add(&nb->b, len);
+			} else {
+				nb = mcumgr_buf_alloc();
+				if (!nb) {
+					LOG_ERR("Failed to allocate mcumgr buffer");
+					/* No free space, drop smp frame */
+					continue;
+				}
+				net_buf_add_mem(nb, conf->recv_buffer, len);
```

Comment:
IIUC, the biggest changes here is the 1) Allocating before the `recvfrom`, and 2) using `simple_add` when you can, and that allows you to prevent a memcpy?

Reply:
Aw yes you also included the `nb ? (char *)net_buf_simple_tail(&nb->b) : conf->recv_buffer` logic in the `recvfrom` call. That makes sense now.

Reply:
Right, `net_buf_simple_add()` merely moves the "tail" pointer inside the `struct net_buf` while `net_buf_add_mem()` copies data from an external buffer _and_ moves the pointer.